File: frontend/src/types/template.ts
/**
 * VM Template types
 */

export interface TemplateCompute {
  cpu_cores: number;
  vcpu?: number;
  cpu_sockets: number;
  cpu_threads: number;
  memory: string;
}

export interface TemplateDisk {
  size: string;
  storage_class?: string;
}

export interface TemplateNetwork {
  type: 'default' | 'multus' | 'bridge';
  multus_network?: string;
}

export interface TemplateCloudInit {
  user_data?: string;
  network_data?: string;
}

export interface TemplateConsole {
  vnc_enabled: boolean;
  serial_console_enabled: boolean;
}

export interface VMTemplate {
  name: string;
  display_name: string;
  description?: string;
  icon?: string;
  category: string;
  os_type: string;
  golden_image_name: string;
  golden_image_namespace: string;
  compute: TemplateCompute;
  disk: TemplateDisk;
  network: TemplateNetwork;
  cloud_init?: TemplateCloudInit;
  console?: TemplateConsole;
  created?: string;
}

export interface VMTemplateCreate {
  name: string;
  display_name: string;
  description?: string;
  icon?: string;
  category?: string;
  os_type?: string;
  golden_image_name: string;
  golden_image_namespace?: string;
  compute?: Partial<TemplateCompute>;
  disk?: Partial<TemplateDisk>;
  network?: Partial<TemplateNetwork>;
  cloud_init?: TemplateCloudInit;
  console?: Partial<TemplateConsole>;
}

export interface VMTemplateListResponse {
  items: VMTemplate[];
  total: number;
}

/**
 * Golden Image types
 */

export type DiskType = 'image' | 'data';

export type ImageScope = 'environment' | 'project';

export interface GoldenImage {
  name: string;
  namespace: string;
  display_name?: string;
  description?: string;
  os_type?: string;
  os_version?: string;
  size?: string;
  status: 'Ready' | 'InUse' | 'Pending' | 'Error' | string;
  error_message?: string;
  source_url?: string;
  created?: string;
  used_by?: string[];  // List of VMs using this image
  disk_type: DiskType;  // image or data
  persistent: boolean;  // If true, disk is not cloned
  scope: ImageScope;  // environment (single ns) or project (all envs)
  project?: string;  // Project name (for project-scoped images)
  environment?: string;  // Environment name (from namespace label)
  // Where this row came from. "cluster" is a DataVolume that exists; "catalog"
  // is a Harbor artifact that has not been materialised into a disk yet.
  // Optional here (the backend always sends it, defaulting to "cluster") so
  // it stays compatible with anything that builds a GoldenImage by hand.
  origin?: 'cluster' | 'catalog';
  // "<project>/<repository>:<tag>" when the row has a Harbor counterpart —
  // present on catalog rows and on cluster rows imported from Harbor.
  catalog_ref?: string | null;
}

export interface GoldenImageCreate {
  name?: string;
  display_name?: string;
  description?: string;
  os_type?: string;
  os_version?: string;
  source_url?: string;
  source_registry?: string;
  source_pvc?: string;  // For cloning existing disk
  source_pvc_namespace?: string;
  size?: string;
  storage_class?: string;
  disk_type?: DiskType;  // image or data
  persistent?: boolean;  // If true, disk is not cloned
  scope?: ImageScope;  // environment (default) or project
  project?: string;  // Project name (when scope=project)
}

export interface GoldenImageUpdate {
  scope?: ImageScope;
  display_name?: string;
  description?: string;
}

export interface GoldenImageListResponse {
  items: GoldenImage[];
  total: number;
  // False when the catalogue could not be read (Harbor down, or the caller's
  // token was rejected). The cluster rows above are still correct and
  // complete — this is only a signal to show a non-blocking warning.
  // Optional/defaulted true so a response from before this flag existed is
  // still read as "the catalogue is fine".
  catalog_available?: boolean;
}

export interface CreateImageFromDiskRequest {
  source_disk_name: string;
  source_namespace: string;
  name?: string;
  display_name?: string;
  description?: string;
  os_type?: string;
  os_version?: string;
}

/**
 * Persistent Disk types
 */

export interface PersistentDisk {
  name: string;
  display_name?: string;
  namespace: string;
  size: string;
  storage_class?: string;
  status: string;
  /** How far an import has got, as CDI reports it ("89.84%"). */
  progress?: string | null;
  attached_to?: string;
  created?: string;
}

export interface PersistentDiskCreate {
  display_name: string;
  size?: string;
  storage_class?: string;
  source_golden_image?: string;
  source_golden_image_namespace?: string;
}

export interface PersistentDiskListResponse {
  items: PersistentDisk[];
  total: number;
}

export interface AttachDiskRequest {
  disk_name: string;
  vm_name: string;
  hotplug?: boolean;
}

/**
 * Network configuration for VM creation
 */

export interface VMNetworkRequest {
  subnet: string;
  static_ip?: string;
}

/**
 * VM from Template request
 */

export interface VMFromTemplateRequest {
  display_name: string;
  template_name: string;
  cpu_cores?: number;
  memory?: string;
  disk_size?: string;
  /** Where the VM's disk goes. Omitted = cluster default, never the template's class. */
  storage_class?: string;
  ssh_key?: string;
  password?: string;
  user_data?: string;
  start?: boolean;
  network?: VMNetworkRequest;
  networks?: VMNetworkRequest[];
}

File: backend/app/models/template.py
"""VM Template models."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class TemplateCompute(BaseModel):
    """Compute specifications for a VM template."""

    cpu_cores: int = Field(2, ge=1, le=128, description="Number of CPU cores")
    cpu_sockets: int = Field(1, ge=1, le=8, description="Number of CPU sockets")
    cpu_threads: int = Field(1, ge=1, le=2, description="Threads per core")
    memory: str = Field("4Gi", description="Memory size (e.g., 4Gi, 8Gi)")


class TemplateDisk(BaseModel):
    """Disk configuration for a VM template."""

    size: str = Field("50Gi", description="Disk size (e.g., 50Gi, 100Gi)")
    storage_class: str | None = Field(None, description="StorageClass name (optional)")


class TemplateNetwork(BaseModel):
    """Network configuration for a VM template."""

    type: Literal["default", "multus", "bridge"] = Field(
        "default", description="Network type"
    )
    multus_network: str | None = Field(
        None, description="Multus network name (for type=multus)"
    )


class TemplateCloudInit(BaseModel):
    """Cloud-init configuration for a VM template."""

    user_data: str | None = Field(None, description="Cloud-init user-data (YAML)")
    network_data: str | None = Field(None, description="Cloud-init network-data (YAML)")


class TemplateConsole(BaseModel):
    """Console configuration for a VM template."""

    vnc_enabled: bool = Field(True, description="Enable VNC console (default: true)")
    serial_console_enabled: bool = Field(False, description="Enable serial console (default: false)")


class VMTemplate(BaseModel):
    """VM Template model."""

    name: str = Field(..., description="Template name (Kubernetes resource name)")
    display_name: str = Field(..., description="Human-readable template name")
    description: str | None = Field(None, description="Template description")
    icon: str | None = Field(None, description="Icon name (e.g., ubuntu, centos, windows)")
    category: str = Field("linux", description="Template category")
    os_type: str = Field("linux", description="OS type (linux, windows)")
    
    # Source golden image
    golden_image_name: str = Field(..., description="Golden image DataVolume name")
    golden_image_namespace: str = Field(
        "golden-images", description="Golden image namespace"
    )
    
    # VM specifications
    compute: TemplateCompute = Field(default_factory=TemplateCompute)
    disk: TemplateDisk = Field(default_factory=TemplateDisk)
    network: TemplateNetwork = Field(default_factory=TemplateNetwork)
    cloud_init: TemplateCloudInit | None = Field(None)
    console: TemplateConsole = Field(default_factory=TemplateConsole)
    
    # Metadata
    created: datetime | None = None
    labels: dict[str, str] = Field(default_factory=dict)
    annotations: dict[str, str] = Field(default_factory=dict)


class VMTemplateCreate(BaseModel):
    """Model for creating a VM template."""

    name: str = Field(
        ...,
        pattern=r"^[a-z0-9]([-a-z0-9]*[a-z0-9])?$",
        max_length=63,
        description="Template name",
    )
    display_name: str = Field(..., max_length=128)
    description: str | None = Field(None, max_length=512)
    icon: str | None = None
    category: str = "linux"
    os_type: str = "linux"
    
    golden_image_name: str
    golden_image_namespace: str  # Must be explicitly specified - same as project namespace
    
    compute: TemplateCompute = Field(default_factory=TemplateCompute)
    disk: TemplateDisk = Field(default_factory=TemplateDisk)
    network: TemplateNetwork = Field(default_factory=TemplateNetwork)
    cloud_init: TemplateCloudInit | None = None
    console: TemplateConsole = Field(default_factory=TemplateConsole)


class VMTemplateUpdate(BaseModel):
    """Model for updating a VM template."""

    display_name: str | None = None
    description: str | None = None
    icon: str | None = None
    category: str | None = None
    
    compute: TemplateCompute | None = None
    disk: TemplateDisk | None = None
    network: TemplateNetwork | None = None
    cloud_init: TemplateCloudInit | None = None
    console: TemplateConsole | None = None


class VMTemplateListResponse(BaseModel):
    """Response model for template list."""

    items: list[VMTemplate]
    total: int


# Golden Image models

class VMImage(BaseModel):
    """VM Image model (DataVolume for VM disks)."""

    name: str
    namespace: str
    display_name: str | None = None
    description: str | None = None
    os_type: str | None = None
    os_version: str | None = None
    size: str | None = None
    status: str  # Ready, InUse, Pending, Error
    error_message: str | None = None  # Error details from DV conditions
    source_url: str | None = None
    created: datetime | None = None
    used_by: list[str] | None = None  # List of VMs using this image
    disk_type: str = "image"  # image or data
    persistent: bool = False  # If true, disk is not cloned
    scope: str = "environment"  # "environment" (single ns) or "project" (all envs)
    project: str | None = None  # Project name (set for project-scoped images)
    environment: str | None = None  # Environment name (from namespace label)


class VMImageCreate(BaseModel):
    """Model for creating a VM image.
    
    If `name` is not provided, it will be auto-generated from `display_name`
    using the pattern {slug}-{uuid6} (e.g. "ubuntu-24-04-server-a7f3e2").
    """

    name: str | None = Field(
        None,
        pattern=r"^[a-z0-9]([-a-z0-9]*[a-z0-9])?$",
        max_length=63,
        description="K8s-safe name. Auto-generated from display_name if not provided.",
    )
    display_name: str | None = Field(None, description="Human-friendly name (stored in annotation)")
    description: str | None = None
    os_type: str = "linux"
    os_version: str | None = None
    
    # Source - one of these
    source_url: str | None = Field(None, description="HTTP URL to download image")
    source_registry: str | None = Field(None, description="Container registry URL")
    source_pvc: str | None = Field(None, description="PVC name to clone from")
    source_pvc_namespace: str | None = Field(None, description="PVC namespace to clone from")
    
    size: str = Field("10Gi", description="Disk size")
    storage_class: str | None = None
    
    # Disk type and persistence
    disk_type: str = Field("image", description="Disk type: image or data")
    persistent: bool = Field(False, description="If true, disk is not cloned")
    
    # Scope: environment (single ns) or project (all envs in project)
    scope: str = Field("environment", description="'environment' or 'project'")
    project: str | None = Field(None, description="Project name (required when scope=project)")


class VMImageListResponse(BaseModel):
    """Response model for VM image list."""

    items: list[VMImage]
    total: int


class VMImageUpdate(BaseModel):
    """Model for updating a VM image (scope, display name, etc.)."""

    scope: str | None = Field(None, description="'environment' or 'project'")
    display_name: str | None = None
    description: str | None = None


# Aliases for backward compatibility
GoldenImage = VMImage
GoldenImageCreate = VMImageCreate
GoldenImageListResponse = VMImageListResponse
GoldenImageUpdate = VMImageUpdate


# Persistent Disk models

class PersistentDisk(BaseModel):
    """Persistent disk model (independent DataVolume)."""

    name: str
    namespace: str
    size: str
    storage_class: str | None = None
    status: str
    attached_to: str | None = None  # VM name if attached
    created: datetime | None = None


class PersistentDiskCreate(BaseModel):
    """Model for creating a persistent disk."""

    name: str = Field(
        ...,
        pattern=r"^[a-z0-9]([-a-z0-9]*[a-z0-9])?$",
        max_length=63,
    )
    size: str = Field("50Gi")
    storage_class: str | None = None
    
    # Optional: clone from image in the same namespace
    source_image: str | None = None


class PersistentDiskListResponse(BaseModel):
    """Response model for persistent disk list."""

    items: list[PersistentDisk]
    total: int


class AttachDiskRequest(BaseModel):
    """Request to attach a disk to a VM."""

    disk_name: str
    vm_name: str
    hotplug: bool = False  # If true, attach without stopping VM


class CreateImageFromDiskRequest(BaseModel):
    """Request to create an image from an existing disk into a project namespace."""

    source_disk_name: str
    source_namespace: str
    
    # Target namespace for the new image (defaults to source_namespace if not provided)
    target_namespace: str | None = None
    
    # New image details
    name: str = Field(
        ...,
        pattern=r"^[a-z0-9]([-a-z0-9]*[a-z0-9])?$",
        max_length=63,
    )
    display_name: str | None = None
    description: str | None = None
    os_type: str = "linux"
    os_version: str | None = None
